jms-lib/src/main/java/com/streamsets/pipeline/lib/jms/config/BaseJmsConfig.java

```diff
@@ -64,6 +64,16 @@
   @ValueChooserModel(DestinationTypeChooserValues.class)
   public DestinationType destinationType = DestinationType.UNKNOWN; // NOTE: same as above
 
+  @ConfigDef(
+      required = false,
+      type = ConfigDef.Type.STRING,
+      label = "Client ID",
+      description = "If specified, must be a unique string used by a single JMS connection instance. Required for Durable Topic Subscriptions",
+      displayPosition = 75,
+      group = "JMS"
+  )
+  public String clientID = null;
```

Comment:
I'm brainstorming here, so bear with me. What if we would add a checkbox "Use Client ID" that would enable or disable this config? Then for JMS systems that do require it the user can primitively select the checkbox and specify it.

And this can be independent on the durable subscriber since as you're saying, not all systems require the client id to be explicitly specified for durable subscriber either.

Reply:
I think that could work. 

To be clear, though, the client ID is not necessarily required for future connections, but in order for the client to use the durable subscription and get messages that were published while they were offline, the `clientID` *must* be the same as it was when the subscription was created. So leaving it to auto-generated client IDs is kind of risky.

Reply:
I will play with a checkbox for clientID later today

Reply:
https://memegenerator.net/img/instances/65418174/just-when-i-thought-i-was-outthey-pull-me-back-in.jpg

Reply:
:-) I submitted the patch to our internal Gerrit to see if things will go well and will wait on the checkbox to actually merge.

Reply:
@jarcec what is the process for upgrade? For instance, if I say that the Boolean for `use clientID` or `durable subscription` are `required`, when I use the new stage library version with an existing pipeline, I get an error that says the required field has no value. Is this how you have done additive properties in the past? Or should the Boolean values be `required = false`?

Reply:
You need to bump the version of the stage here:

https://github.com/streamsets/datacollector/blob/master/jms-lib/src/main/java/com/streamsets/pipeline/stage/origin/jms/JmsDSource.java#L36

And create proper upgrader here:

https://github.com/streamsets/datacollector/blob/master/jms-lib/src/main/java/com/streamsets/pipeline/stage/origin/jms/JmsSourceUpgrader.java#L29

Reply:
I am going to be an old man before this PR is done...

Reply:
We're almost there :)

Reply:
so for the upgrader, I just pick the next increment? so 7?

Reply:
Correct and then upgrade from 6 to 7 (e.g. adding all the new configuration options).

Reply:
```suggestion
switch(context.getFromVersion()) {
      case 1:
        upgradeV1ToV2(configs);
        if (context.getToVersion() == 2) {
          break;
        }
        // fall through
      case 2:
        upgradeV2ToV3(configs);
        if (context.getToVersion() == 3) {
          break;
        }
        // fall through
      case 3:
        upgradeV3ToV4(configs);
        if (context.getToVersion() == 4) {
          break;
        }
        // fall through
      case 4:
        if (context.getToVersion() == 5) {
          break;
        }
        upgradeV4ToV5(configs);
      // fall through
      case 5:
        upgradeV5ToV6(configs, context);
        break;
      default:
        throw new IllegalStateException(Utils.format("Unexpected fromVersion {}", context.getFromVersion()));
}
```
why in `case 4:` is the test for the `getToVersion` done first and then the upgrade function vs. all the others where the upgrade function is done and then the `getToVersion` is checked?

Reply:
I believe that that is a bug @mackenza , and a fairly old one. The case 4 is correct, the others not so much.

Reply:
am I missing something here? In `case 4` if I am upgrading from 4 (and thus match the case) to 5, it's going to `break;` and not do the `upgradeV4ToV5()`.  Again, I may be missing something. 

The only reason I ask is I need to add something like that for my `upgradeV6ToV7()` I need to write and I need to alter the `case 5:` to do that fall through thing, don't I?

Reply:
Yeah we should upgrade and then break, sorry.

Reply:
ok I added the upgrade stuff and also the boolean for entering clientID. I think it's all set now.

Reply:
Thanks @mackenza ! Let me upload the latest variant to gerrit for automated testing and will merge if successful!